### PR TITLE
Implement JDK8 Optional methods on the internal Optional class

### PR DESCRIPTION
**Pull request by vinaysshenoy**
https://www.pivotaltracker.com/story/show/173238805

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/util/Optional.kt [hunk 1]
```diff
@@ -40,36 +74,46 @@ sealed class Optional<out T : Any> {
...
+
+    fun <T> ofNullable(value: T?): Optional<T> = if (value != null) Just(value) else None()
   }
 }
 
-data class Just<out T : Any>(val value: T) : Optional<T>() {
-  override fun toString() = "Just($value)"
-  override fun toNullable(): T = value
-}
+class Just<T>(val value: T) : Optional<T>(JOptional.of(value)) {
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/util/Optional.kt]: I think was removed in my first commit, which was reordered later. Can you check the order of commits locally?

File: app/src/main/java/org/simple/clinic/util/Optional.kt [hunk 2]
```diff
@@ -40,36 +74,46 @@ sealed class Optional<out T : Any> {
...
+      other == null || other !is Just<*> -> false
+      else -> other.value == value
+    }
+  }
 
-  override fun component1(): Nothing? = null
+  override fun hashCode(): Int = value.hashCode()
+}
 
+class None<T> : Optional<T>(JOptional.empty()) {
```
[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/util/Optional.kt]: Same comment as earlier.

File: app/src/main/java/org/simple/clinic/util/Optional.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/util/Optional.kt]: I have checked it in this commit bec2f52efb2a8dceee0440c36256cecf1a10fed2, this shows as first commit locally.

[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/util/Optional.kt]: Ah, I see the problem. Will fix.

[the remark this answers is not in the document]
Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/util/Optional.kt]: I tried to fix the commits individually, but the commits were getting too hairy to resolve conflicts correctly. I ended up squashing the first two commits together.

**Pull request merged**